### DISCUSSION TITLE
How to enable VAD

### DISCUSSION THREAD

**Discussion by cfry** [speechly/speechly#263] [Q&A] [answered]
I want to have Speechly automatically detect when I'm speaking, and send such utterances
to be recognized.
On https://speechly.github.io/speechly-unity-dotnet/api/Speechly.SLUClient.SpeechlyClient.html#Speechly_SLUClient_SpeechlyClient_OnSegmentChange
I quote:
OnSegmentChange
...
You can control when to start and stop process speech either manually with Start(String) and Stop(). Alternatively, you may let SpeechlyClient control this automatically by setting setting AudioProcessorOptions.VADControlsListening to true upon SpeechlyClient.Initialize() or by calling SpeechlyClient.AdjustAudioProcessor(true).

So I think this is what I want. 
I'm using JavaScript, so as always, this "language agnostic, example free doc" makes
figuring out exact syntax difficult.
If I call:
SpeechlyClient.AdjustAudioProcessor(true)
I get an error saying AdjustAudioProcessor is not a method.

Then  " setting AudioProcessorOptions.VADControlsListening to true "
implies  setting AudioProcessorOptions.VADControlsListening = true 
but AudioProcessorOptions isn't bound globally so that part isn't useful either.

So I tried to figure out another way of doing this,
which I'm guessing I can do in the BrowserClient instance constructor in syntax like:
let client = new BrowserClient({
                    appId: 'xxx',
                    vad: {enabled: false, noiseGateDb: -24.0} 
But I see no examples in the doc.
Would just changing "false" to "true" above do it?

Once I do this, there's no need to call client.start()  right?

If I want to "pause" the listening, should I call client.stop()
and then to start up again call client.start() ?

Suppose I get VAD enabled. What if an utterance I make starts off softly, below
the energy threshold, then ramps up to above the threshold?
Is that soft initial sound cut off so that reco won't be so good,
or do you continuously record anyway an initial 250ms or so in a circular buffer so that
you can then pass in that initial sound to improve reco accuracy?
(though this is a good technique, I'm worried it might eat up processor
bandwidth and degrade my app performance)

**Comment by teelisyys** [comment, marked as answer]
Hi @cfry!

The documentation you found is the reference four our Unity client. As you found out, the browser-client uses different syntax. Changing the `false` to `true` in the VAD options will indeed enable the VAD in the browser-client. Once initially enabled, you can toggle VAD on/off on-the-fly like so:

`client.adjustAudioProcessor( { vad: { enabled: true }})` and `client.adjustAudioProcessor( { vad: { enabled: false }})`

Setting `enabled: false` also ensures audio processing is stopped.

Note that when VAD is enabled, it takes control of calling `BrowserClient.start()` and `stop()`. The effect of manually calling these with VAD enabled is currently unspecified.

* * *

VAD activates when signal energy exceeds both the absolute energy threshold and the dynamic signal-to-noise threshold for a period of time. This allows adjusting to changes in the signal strength. These thresholds can be adjusted with [the options](https://github.com/speechly/speechly/blob/main/libraries/browser-client/docs/interfaces/client.VadOptions.md).

The browser-client VAD will have a CPU cost, but will reduce the network traffic if the input audio contains silence or other low energy parts. This is a trade-off and in a very limited CPU environment might not be worth making. Those interested in the implementation details can find the implementation [here](https://github.com/speechly/speechly/blob/main/libraries/browser-client/src/audioprocessing/EnergyThresholdVAD.ts).

An example for type script can be found in this repo from the [browser-client-example](https://github.com/speechly/speechly/blob/main/examples/browser-client-example/src/index.ts#L32) and for plain javascript in our [blog](https://www.speechly.com/blog/create-a-webrtc-video-chat-app-with-speechly-transcription).
